### PR TITLE
Merge Request from Dev-Beta Branch

### PR DESCRIPTION
**Pull request by ilBarlo**

### CODE REVIEW

File: cmd/doc.go [hunk 1]
```diff
@@ -0,0 +1,2 @@
+// Package main is where everything starts.
+package main
```
Comment (LucaRocco97) [inline cmd/doc.go]: Why do you need it?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,8 @@
-# FlavourGeneratorConsumer
```
Comment (LucaRocco97) [inline README.md:1]: Leave a blank line below

File: README.md [hunk 3]
```diff
@@ -1,2 +1,8 @@
...
-Consumer for the FlavourGenerator, who retrieves the metrics from the other node via RabbitMQ
+# FlavourGenerator - Consumer
```
Comment (LucaRocco97) [inline README.md:1]: Leave a blank line below

File: README.md [hunk 4]
```diff
@@ -1,2 +1,8 @@
...
+This repository contains a simple Golang application that acts as a receiver for messages published on a RabbitMQ channel. The application is built using the AMQP client library and listens to the specified RabbitMQ queue for incoming messages. The received messages are then logged to the console for the user's convenience.
+
+To use this repository, the following steps need to be taken:
+1. Set up a RabbitMQ server.
+
+## About
```
Comment (LucaRocco97) [inline README.md:9]: Leave a blank line below

File: README.md [hunk 5]
```diff
@@ -1,2 +1,8 @@
...
+For the repo of the producer, click this [link](https://github.com/ilBarlo/FlavourGeneratorProducer)
```
Comment (LucaRocco97) [inline README.md:11]: Leave a blank line below

File: pkg/flavourmanager/consumer.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	"fmt"
+	"log"
+)
+
+// Function that start a new Consumer on a RabbitMQ Channel
+func StartConsumer(queueName string, url string) error {
+
+	conn, ch, err := createChannel(url)
+	if err != nil {
+		return fmt.Errorf("failed to create channel: %v", err)
```
Comment (LucaRocco97) [inline pkg/flavourmanager/consumer.go]: Use %w instad of %v to wrap errors

File: pkg/flavourmanager/consumer.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	}
+	defer ch.Close()
+	defer conn.Close()
+
+	err = declareQueue(ch, queueName)
+	if err != nil {
+		return fmt.Errorf("failed to declare queue: %v", err)
```
Comment (LucaRocco97) [inline pkg/flavourmanager/consumer.go]: use %w instead of %v to wrap errors

File: pkg/flavourmanager/consumer.go [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+	}
+
+	msgs, err := consumeMsgs(ch, queueName)
+	if err != nil {
+		return fmt.Errorf("failed to consume messages: %v", err)
```
Comment (LucaRocco97) [inline pkg/flavourmanager/consumer.go]: Use %w instad of %v to wrap errors

File: pkg/flavourmanager/consumer.go [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+	}
+
+	forever := make(chan bool)
+	go func() {
+		for d := range msgs {
+			handleMsg(d.Body)
+		}
+	}()
+
+	log.Printf(" [*] Waiting for messages. To exit press CTRL+C")
```
Comment (LucaRocco97) [inline pkg/flavourmanager/consumer.go]: This print is ok when you run the consumer on your computer but remember that the consumer will be ran in a container (or in a server) so you won't be have the possibility to stop it by using CTRL+C

File: pkg/flavourmanager/models.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
+package flavourmanager
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// Info of the node
```
Comment (LucaRocco97) [inline pkg/flavourmanager/models.go]: I'd prefer that each struct's comment starts with the struct's name (e.g. // NodeInfo ...)

File: pkg/flavourmanager/models.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+type NodeInfo struct {
+	UID             string          `json:"uid"`
+	Name            string          `json:"name"`
+	Architecture    string          `json:"architecture"`
+	OperatingSystem string          `json:"os"`
+	ResourceMetrics ResourceMetrics `json:"resources"`
+}
+
+// Metrics of the resource of a certain node
```
Comment (LucaRocco97) [inline pkg/flavourmanager/models.go]: I'd prefer that each struct's comment starts with the struct's name (e.g. // ResourceMetrics ...)

File: pkg/flavourmanager/models.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+	CPUAvailable    string `json:"availableCPU"`
+	MemoryTotal     string `json:"totalMemory"`
+	MemoryAvailable string `json:"availableMemory"`
+}
+
+// Map of the nodes
+type NodeMap map[string]NodeInfo
+
+// Update the NodeMap
+func updateNodeMap(nodeMap NodeMap, newNode NodeInfo) {
```
Comment (LucaRocco97) [inline pkg/flavourmanager/models.go:29]: I'd prefer that each function's comment starts with the function's name (e.g. // updateNodeMap ...)

File: pkg/flavourmanager/models.go [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+// Metrics of the resource of a certain node
+type ResourceMetrics struct {
+	CPUTotal        string `json:"totalCPU"`
+	CPUAvailable    string `json:"availableCPU"`
+	MemoryTotal     string `json:"totalMemory"`
+	MemoryAvailable string `json:"availableMemory"`
+}
+
+// Map of the nodes
```
Comment (LucaRocco97) [inline pkg/flavourmanager/models.go]: I'd prefer that each var's comment starts with the var's name (e.g. // NodeMap ...)

File: pkg/flavourmanager/models.go [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+type NodeMap map[string]NodeInfo
+
+// Update the NodeMap
+func updateNodeMap(nodeMap NodeMap, newNode NodeInfo) {
+	if existingNode, ok := nodeMap[newNode.UID]; ok {
+		// The node already exists, check if there is some changes
+		if ok && reflect.DeepEqual(newNode, existingNode) {
+			// Nothing
```
Comment (LucaRocco97) [inline pkg/flavourmanager/models.go]: What? Fix it immediately! Just negate the condition and put the instructions in the else branch inside the if branch

File: pkg/flavourmanager/rabbitmq_manager.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+	"encoding/json"
+	"fmt"
+	"log"
+
+	"github.com/streadway/amqp"
+)
+
+var nodeMap = make(NodeMap)
+
+// Function to create RabbitMQ Channel
```
Comment (LucaRocco97) [inline pkg/flavourmanager/rabbitmq_manager.go]: I'd prefer that each function's comment starts with the function's name (e.g. // createChannel ...)

File: pkg/flavourmanager/rabbitmq_manager.go [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+	}
+	// Create Channel
+	ch, err := conn.Channel()
+	if err != nil {
+		return nil, nil, err
+	}
+	return conn, ch, nil
+}
+
+// Function to declare a queue on a RabbitMQ Channel
```
Comment (LucaRocco97) [inline pkg/flavourmanager/rabbitmq_manager.go]: I'd prefer that each function's comment starts with the function's name (e.g. // declareQueue ...)

File: pkg/flavourmanager/rabbitmq_manager.go [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+		false,     // no-wait
+		nil,       // arguments
+	)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Function to consume messages arrived on a RabbitMQ Channel
```
Comment (LucaRocco97) [inline pkg/flavourmanager/rabbitmq_manager.go]: I'd prefer that each function's comment starts with the function's name (e.g. // consumeMsgs ...)

File: pkg/flavourmanager/rabbitmq_manager.go [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+		nil,   // arguments
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to register a consumer: %w", err)
+	}
+
+	return msgs, nil
+}
+
+// Function that handleMsg
```
Comment (LucaRocco97) [inline pkg/flavourmanager/rabbitmq_manager.go]: I'd prefer that each function's comment starts with the function's name (e.g. // handleMsg ...)

**Review by LucaRocco97 — COMMENTED** [review]
Generally speaking it is ok but minor issues are present

File: cmd/doc.go (hunk 1 above)
Comment (ilBarlo) [reply cmd/doc.go]: Done.

File: README.md (hunk 3 above)
Comment (ilBarlo) [reply README.md:1]: Done.

File: pkg/flavourmanager/consumer.go (hunk 6 above)
Comment (ilBarlo) [reply pkg/flavourmanager/consumer.go]: Done.

File: pkg/flavourmanager/consumer.go (hunk 9 above)
Comment (ilBarlo) [reply pkg/flavourmanager/consumer.go]: Done.

File: pkg/flavourmanager/models.go (hunk 14 above)
Comment (ilBarlo) [reply pkg/flavourmanager/models.go]: Done.

**Comment by ilBarlo** [comment]
Pull request #1 from Dev-Beta Branch. Code reviewed by Luca.

**Pull request merged**